<a href="https://colab.research.google.com/github/ChonChanokAng/BSC_DPDM23/blob/main/Freq_HW1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import os

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [5]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_dpdm23_data/reduced_marketbasket.csv'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [6]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/BSC_dpdm23_data/reduced_marketbasket.csv'),encoding='latin-1')
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [7]:
data #chek data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [22]:
data.groupby(['Country']).sum() #ดูข้อมูลมาจากกี่ประเทศ

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
<ipython-input-22-25f447ef123a>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  data.groupby(['Country']).sum() #ดูข้อมูลมาจากกี่ประเทศ


Quantity    UnitPrice    CustomerID
Country                                                  
Australia                83653     4054.750  1.569300e+07
Austria                   4827     1701.520  5.021102e+06
Bahrain                    260       86.570  2.100270e+05
Belgium                  23152     7540.130  2.571829e+07
Brazil                     356      142.600  4.086080e+05
Canada                    2763      910.580  2.615483e+06
Channel Islands           9479     3738.550  1.128522e+07
Cyprus                    6317     3920.070  7.715880e+06
Czech Republic             592       88.150  3.834300e+05
Denmark                   8188     1266.950  4.876734e+06
EIRE                    142637    48447.190  1.103917e+08
European Community         497      294.050  9.215880e+05
Finland                  10666     3786.850  8.699324e+06
France                  110480    43031.990  1.076489e+08
Germany                 117448    37666.000  1.200751e+08
Greece                    1556      713.290  2.008584e+06
Hong Kong                 4769    12241.500  0.000000e+00
Iceland                   2458      481.210  2.247154e+06
Israel                    4353     1079.040  3.164467e+06
Italy                     7999     3879.390  1.015666e+07
Japan                    25218      814.860  4.567292e+06
Lebanon                    386      242.440  5.743800e+05
Lithuania                  652       99.440  5.366200e+05
Malta                      944      666.010  2.158496e+06
Netherlands             200128     6492.550  3.419054e+07
Norway                   19247     6529.060  1.350765e+07
Poland                    3653     1422.270  4.341972e+06
Portugal                 16180    13037.540  1.886480e+07
RSA                        352      248.100  7.218680e+05
Saudi Arabia                75       24.110  1.256500e+05
Singapore                 5234    25108.890  2.918376e+06
Spain                    26824    12633.450  3.268929e+07
Sweden                   35637     1806.830  6.790083e+06
Switzerland              30325     6813.690  2.377592e+07
USA                       1034      644.980  3.672086e+06
United Arab Emirates       982      229.890  1.018952e+06
United Kingdom         4263829  2245715.474  5.626433e+09
Unspecified               3300     1204.010  3.348046e+06

>>แปลงdata ให้อยู่ในรูปtransaction

In [8]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [9]:
tarnsac = data.groupby('InvoiceNo')['StockCode'].apply(list)

>>Association Rules

In [10]:
!pip install apyori
from apyori import apriori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=6b58630dcce216960027a4780d238f1da51b7551d1a1bf72fb42307c32f9fc84
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


#Hw1 หา Frequent set item set และ Association rules


*   ของช่วงเวลา Christmas & New year (1 dec - 30 dec)
*   ของประเทศยุโรป
*   ของช่วงเวลาตรุษจีนในทวีป Asia ( 15 Jan - 15 Feb)
*   ของช่วงเวลา Christmas & New year Europe (1 dec - 30 dec)
*   Valentine's day
*   โดยจัดกลุ่มตามประเทศ
groupby ประเทศ - เดือน

In [11]:
import datetime
import pandas as pd
import os
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [12]:
data['InvoiceDate'] #chek InvoiceDate

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0         12/01/2010 08:26
1         12/01/2010 08:26
2         12/01/2010 08:26
3         12/01/2010 08:26
4         12/01/2010 08:26
                ...       
541904    12/09/2011 12:50
541905    12/09/2011 12:50
541906    12/09/2011 12:50
541907    12/09/2011 12:50
541908    12/09/2011 12:50
Name: InvoiceDate, Length: 541909, dtype: object

In [13]:
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%Y %H:%M')

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [14]:
data['Date']=data['InvoiceDate'].dt.date

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [15]:
data['Date']=pd.to_datetime(data.Date, format='%Y/%m/%d')

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [16]:
data

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country       Date  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom 2010-12-01  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom 2010-12-01  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
...                    ...        ...         ...             ...        ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France 2011-12-09  
541905 2011-12-09 12:50:00       2.10     12680.0          France 2011-12-09  
541906 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541907 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541908 2011-12-09 12:50:00       4.95     12680.0          France 2011-12-09  

[541909 rows x 9 columns]

## ของช่วงเวลา Christmas & New year (1 dec - 30 dec)

In [17]:
#Select 1-31 Dec
Dec= data[(data['Date'].dt.month == 12) & (data['Date'].dt.day >= 1) & (data['Date'].dt.day <= 31)]

#printcheck
print(Dec)


       InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country       Dat

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [18]:
Dec

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country       Date  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom 2010-12-01  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom 2010-12-01  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
...                    ...        ...         ...             ...        ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France 2011-12-09  
541905 2011-12-09 12:50:00       2.10     12680.0          France 2011-12-09  
541906 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541907 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541908 2011-12-09 12:50:00       4.95     12680.0          France 2011-12-09  

[68006 rows x 9 columns]

In [19]:
transac1 = Dec.groupby('InvoiceNo')['Description'].apply(list).reset_index()
print(transac1)

     InvoiceNo                                        Description
0       536365  [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1       536366  [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2       536367  [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3       536368  [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4       536369                         [BATH BUILDING BLOCK WORD]
...        ...                                                ...
3035   C581484                      [PAPER CRAFT , LITTLE BIRDIE]
3036   C581490  [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
3037   C581499                                           [Manual]
3038   C581568                       [VICTORIAN SEWING BOX LARGE]
3039   C581569  [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...

[3040 rows x 2 columns]


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [20]:
transac1apply = transac1['Description'].apply(lambda x: [str(item) for item in x])
transac1apply

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0       [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1       [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2       [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3       [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4                              [BATH BUILDING BLOCK WORD]
                              ...                        
3035                        [PAPER CRAFT , LITTLE BIRDIE]
3036    [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
3037                                             [Manual]
3038                         [VICTORIAN SEWING BOX LARGE]
3039    [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...
Name: Description, Length: 3040, dtype: object

In [27]:
from apyori import apriori

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [28]:
frequentItemset = list(apriori(transac1apply, min_support=0.036))

for record in frequentItemset:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()

yes
Items: ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKELIKE RED 
Support: 0.0368 (3.68%)
Rule 1: (None -> ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKELIKE RED  : Confidence=0.0368, Lift=1.0000)
Rule 2: (ALARM CLOCK BAKELIKE GREEN -> ALARM CLOCK BAKELIKE RED  : Confidence=0.7089, Lift=14.4627)
Rule 3: (ALARM CLOCK BAKELIKE RED  -> ALARM CLOCK BAKELIKE GREEN : Confidence=0.7517, Lift=14.4627)

yes
Items: CHOCOLATE HOT WATER BOTTLE, HOT WATER BOTTLE I AM SO POORLY
Support: 0.0365 (3.65%)
Rule 1: (None -> CHOCOLATE HOT WATER BOTTLE, HOT WATER BOTTLE I AM SO POORLY : Confidence=0.0365, Lift=1.0000)
Rule 2: (CHOCOLATE HOT WATER BOTTLE -> HOT WATER BOTTLE I AM SO POORLY : Confidence=0.5163, Lift=9.7484)
Rule 3: (HOT WATER BOTTLE I AM SO POORLY -> CHOCOLATE HOT WATER BOTTLE : Confidence=0.6894, Lift=9.7484)

yes
Items: PAPER CHAIN KIT VINTAGE CHRISTMAS, PAPER CHAIN KIT 50'S CHRISTMAS 
Support: 0.0477 (4.77%)
Rule 1: (None -> PAPER CHAIN KIT VINTAGE CHRISTMAS, PAPER CHAIN KIT 50'S CHRISTMAS  : 

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


มี 3 itemset ที่ผ่าน 0.036
1. คนที่ซื้อทั้ง PAPER CHAIN KIT 50'S CHRISTMAS และ PAPER CHAIN KIT VINTAGE CHRISTMAS จาก dataset ทั้งหมด มี 4.77%

   คนที่ซื้อ PAPER CHAIN KIT 50'S CHRISTMAS และ PAPER CHAIN KIT VINTAGE CHRISTMAS มี 4.77%
   
   คนที่ซื้อ PAPER CHAIN KIT 50'S CHRISTMAS แล้วค่อยซื้อ PAPER CHAIN KIT VINTAGE CHRISTMAS มี 51.6%

   คนที่ซื้อ PAPER CHAIN KIT VINTAGE CHRISTMAS แล้วค่อยซื้อ PAPER CHAIN KIT 50'S CHRISTMAS มี 66.51%


2. คนที่ซื้อทั้ง ALARM CLOCK BAKELIKE RED และ ALARM CLOCK BAKELIKE GREEN จาก dataset ทั้งหมด มี 3.68%

   คนที่ซื้อ ALARM CLOCK BAKELIKE RED และ ALARM CLOCK BAKELIKE GREEN มี 3.68%

   คนที่ซื้อ ALARM CLOCK BAKELIKE GREEN แล้วค่อยซื้อ ALARM CLOCK BAKELIKE RED มี 75.17%

   คนที่ซื้อ ALARM CLOCK BAKELIKE RED แล้วค่อยซื้อ ALARM CLOCK BAKELIKE GREEN มี 70.89%



3. คนที่ซื้อ CHOCOLATE HOT WATER BOTTLE และ HOT WATER BOTTLE I AM SO POORLY จาก dataset ทั้งหมด มี 3.65%

   คนที่ซื้อ CHOCOLATE HOT WATER BOTTLE และ HOT WATER BOTTLE I AM SO POORLY มี 3.65%

   คนที่ซื้อ CHOCOLATE HOT WATER BOTTLE แล้วค่อยซื้อ HOT WATER BOTTLE I AM SO POORLY มี 51.63%

   คนที่ซื้อ HOT WATER BOTTLE I AM SO POORLY แล้วค่อยซื้อ CHOCOLATE HOT WATER BOTTLE มี 68.94%



มี 3 itemset ที่ผ่าน 0.036

RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,
<t>
<t>

คนที่ซื้อทั้ง 22386, 85099B จาก dataset ทั้งหมดมี 3.21%

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

คนที่ซื้อ 22386 แล้วค่อยซื้อ 85099B มีตั้ง 67.66%


OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),


คนที่ซื้อ 85099B แล้วค่อยซื้อ 22386 มี 39.01%


OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])

## ของประเทศยุโรป

##จัดกลุ่มประเทศตามทวีป

In [24]:
set(data['Country']) #check country

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

In [25]:
Eur = data[(data['Country']!='Bahrain') | (data['Country']!='Hong Kong')| (data['Country']!='Singapore') | (data['Country']!='Japan') | (data['Country']!='Israel') | (data['Country']!='Lebanonl') |
           (data['Country']!='Saudi Arabia') | (data['Country']!='United Arab Emirates')| (data['Country']!='Australia')| (data['Country']!='Brazil') | (data['Country']!='RSA')| (data['Country']!='Canada')|
           (data['Country']!='USA')]
Eur #แบ่งทวีปยุโรป

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country       Date  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom 2010-12-01  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom 2010-12-01  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
...                    ...        ...         ...             ...        ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France 2011-12-09  
541905 2011-12-09 12:50:00       2.10     12680.0          France 2011-12-09  
541906 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541907 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541908 2011-12-09 12:50:00       4.95     12680.0          France 2011-12-09  

[541909 rows x 9 columns]

In [26]:
Asia = data[(data['Country']=='Bahrain') | (data['Country']=='Hong Kong')| (data['Country']=='Singapore') | (data['Country']=='Japan') | (data['Country']=='Israel') |
            (data['Country']=='Lebanon') | (data['Country']=='Saudi Arabia') | (data['Country']=='United Arab Emirates')]
Asia #แบ่งทวีปเอเชีย

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


InvoiceNo StockCode                     Description  Quantity  \
9783      537218     85016   SET OF 6 VINTAGE NOTELETS KIT         6   
9784      537218     21506      FANCY FONT BIRTHDAY CARD,         12   
9785      537218     22716              CARD CIRCUS PARADE        12   
9786      537218     22983             CARD BILLBOARD FONT        12   
9787      537218     22030          SWALLOWS GREETING CARD        12   
...          ...       ...                             ...       ...   
530276   C580832     21731   RED TOADSTOOL LED NIGHT LIGHT        -2   
530277   C580832     23480  MINI LIGHTS WOODLAND MUSHROOMS       -15   
530278   C580832     22466  FAIRY TALE COTTAGE NIGHT LIGHT       -17   
530279   C580832     23084              RABBIT NIGHT LIGHT        -7   
530280   C580832     22621      TRADITIONAL KNITTING NANCY        -1   

               InvoiceDate  UnitPrice  CustomerID Country       Date  
9783   2010-12-05 15:46:00       2.55     12763.0   Japan 2010-12-05  
9784   2010-12-05 15:46:00       0.42     12763.0   Japan 2010-12-05  
9785   2010-12-05 15:46:00       0.42     12763.0   Japan 2010-12-05  
9786   2010-12-05 15:46:00       0.42     12763.0   Japan 2010-12-05  
9787   2010-12-05 15:46:00       0.42     12763.0   Japan 2010-12-05  
...                    ...        ...         ...     ...        ...  
530276 2011-12-06 11:40:00       1.45     12753.0   Japan 2011-12-06  
530277 2011-12-06 11:40:00       3.29     12753.0   Japan 2011-12-06  
530278 2011-12-06 11:40:00       1.65     12753.0   Japan 2011-12-06  
530279 2011-12-06 11:40:00       1.79     12753.0   Japan 2011-12-06  
530280 2011-12-06 11:40:00       1.45     12753.0   Japan 2011-12-06  

[1314 rows x 9 columns]

##Valentine's day

In [29]:
#Select 14 Feb
Feb = data[(data['Date'].dt.month == 2) & (data['Date'].dt.day == 14)]

#printcheck
print(Feb)

      InvoiceNo StockCode                         Description  Quantity  \
88847    543821     22926      IVORY GIANT GARDEN THERMOMETER         2   
88848    543821     22928     YELLOW GIANT GARDEN THERMOMETER         2   
88849    543821     22925       BLUE GIANT GARDEN THERMOMETER         2   
88850    543821     85175               CACTI T-LIGHT CANDLES        16   
88851    543821     22456     NATURAL SLATE CHALKBOARD LARGE          3   
...         ...       ...                                 ...       ...   
89950    543979     21889              WOODEN BOX OF DOMINOES         1   
89951    543979     85176          SEWING SUSAN 21 NEEDLE SET         1   
89952    543979     37446  MINI CAKE STAND WITH HANGING CAKES         1   
89953    543979     21657     MILK BOTTLE WITH GLASS STOPPER          1   
89954   C543980     22728           ALARM CLOCK BAKELIKE PINK        -1   

              InvoiceDate  UnitPrice  CustomerID         Country       Date  
88847 2011-02-14 09:2

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [30]:
Feb

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


InvoiceNo StockCode                         Description  Quantity  \
88847    543821     22926      IVORY GIANT GARDEN THERMOMETER         2   
88848    543821     22928     YELLOW GIANT GARDEN THERMOMETER         2   
88849    543821     22925       BLUE GIANT GARDEN THERMOMETER         2   
88850    543821     85175               CACTI T-LIGHT CANDLES        16   
88851    543821     22456     NATURAL SLATE CHALKBOARD LARGE          3   
...         ...       ...                                 ...       ...   
89950    543979     21889              WOODEN BOX OF DOMINOES         1   
89951    543979     85176          SEWING SUSAN 21 NEEDLE SET         1   
89952    543979     37446  MINI CAKE STAND WITH HANGING CAKES         1   
89953    543979     21657     MILK BOTTLE WITH GLASS STOPPER          1   
89954   C543980     22728           ALARM CLOCK BAKELIKE PINK        -1   

              InvoiceDate  UnitPrice  CustomerID         Country       Date  
88847 2011-02-14 09:24:00       5.95     17667.0  United Kingdom 2011-02-14  
88848 2011-02-14 09:24:00       5.95     17667.0  United Kingdom 2011-02-14  
88849 2011-02-14 09:24:00       5.95     17667.0  United Kingdom 2011-02-14  
88850 2011-02-14 09:24:00       0.42     17667.0  United Kingdom 2011-02-14  
88851 2011-02-14 09:24:00       4.95     17667.0  United Kingdom 2011-02-14  
...                   ...        ...         ...             ...        ...  
89950 2011-02-14 15:43:00       1.25     14606.0  United Kingdom 2011-02-14  
89951 2011-02-14 15:43:00       0.85     14606.0  United Kingdom 2011-02-14  
89952 2011-02-14 15:43:00       1.45     14606.0  United Kingdom 2011-02-14  
89953 2011-02-14 15:43:00       6.95     14606.0  United Kingdom 2011-02-14  
89954 2011-02-14 16:39:00       3.75     14189.0  United Kingdom 2011-02-14  

[1108 rows x 9 columns]

In [45]:
frequentItemset = list(apriori(transac2apply, min_support=0.073))

for record in frequentItemset:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()

yes
Items: SET OF 3 CAKE TINS PANTRY DESIGN , HOMEMADE JAM SCENTED CANDLES
Support: 0.0909 (9.09%)
Rule 1: (None -> HOMEMADE JAM SCENTED CANDLES, SET OF 3 CAKE TINS PANTRY DESIGN  : Confidence=0.0909, Lift=1.0000)
Rule 2: (HOMEMADE JAM SCENTED CANDLES -> SET OF 3 CAKE TINS PANTRY DESIGN  : Confidence=1.0000, Lift=6.8750)
Rule 3: (SET OF 3 CAKE TINS PANTRY DESIGN  -> HOMEMADE JAM SCENTED CANDLES : Confidence=0.6250, Lift=6.8750)

yes
Items: SET OF 6 SPICE TINS PANTRY DESIGN, SET OF 3 CAKE TINS PANTRY DESIGN 
Support: 0.0909 (9.09%)
Rule 1: (None -> SET OF 6 SPICE TINS PANTRY DESIGN, SET OF 3 CAKE TINS PANTRY DESIGN  : Confidence=0.0909, Lift=1.0000)
Rule 2: (SET OF 3 CAKE TINS PANTRY DESIGN  -> SET OF 6 SPICE TINS PANTRY DESIGN : Confidence=0.6250, Lift=5.7292)
Rule 3: (SET OF 6 SPICE TINS PANTRY DESIGN -> SET OF 3 CAKE TINS PANTRY DESIGN  : Confidence=0.8333, Lift=5.7292)



/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
